File: ruby/luhn/luhn.rb
class Luhn
  attr_reader :number

  def initialize(number)
    @number = number.to_s.split("")
  end

  def integer_chars(number)
    number.to_s.split("")
  end

  def addends
    removed = number.shift.to_i if number.length.odd?
    split = number.map.with_index do |digit, index|
      index.even? ? double_and_reduce(digit) : digit.to_i
    end
    split.unshift(removed) if removed
    split
  end

  def double_and_reduce(digit)
    if digit.to_i * 2 > 9
      digit.to_i * 2 - 9
    else
      digit.to_i * 2
    end
  end

  def checksum
    addends.reduce(0, :+)
  end

  def valid?
    checksum % 10 == 0
  end

  def self.create(invalid_number)
    num = invalid_number.to_s.split("")
    (0..9).each do |option|
      new_num = num.push(option.to_s)
      if new(integer(new_num)).valid?
        return integer(num)
      end
      num.pop
    end
  end

  def self.integer(num)
    num.join("").to_i
  end
end

File: python/luhn/luhn.py
class Luhn():
    def __init__(self, sequence):
        self.sequence = sequence

    def addends(self):
        new_sequence = []
        if len(str(self.sequence)) % 2 == 0:
            for index, number in enumerate(str(self.sequence)):
                if index % 2 != 0:
                    new_sequence.append(int(number))
                else:
                    new_sequence.append(self.double_number(number))
        else:
            for index, number in enumerate(str(self.sequence)):
                if index % 2 == 0:
                    new_sequence.append(int(number))
                else:
                    new_sequence.append(self.double_number(number))
        return new_sequence

    def double_number(self, number):
        if int(number) * 2 > 9:
            return int(number) * 2 - 9
        else:
            return int(number) * 2

    def checksum(self):
        addended = int(''.join(map(str, self.addends())))
        return sum(int(digit) for digit in str(addended))

    def is_valid(self):
        return self.checksum() % 10 == 0

    @staticmethod
    def create(number):
        possible_checksums = range(0, 9)
        for possible_checksum in possible_checksums:
            if Luhn(int(str(number) + str(possible_checksum))).is_valid():
                return int(str(number) + str(possible_checksum))
